pkg/controllers/mcs/serviceexport_controller.go

```diff
@@ -0,0 +1,217 @@
+			se, metav1.UpdateOptions{})
+		if err != nil {
+			d := time.Second
+			return &d, err
+		}
+	}
+
+	// recycle corresponding endpoint slice in parent cluster.
+	if seTerminating {
+		if err = r.parentk8sClient.DiscoveryV1().EndpointSlices(r.dedicatedNamespace).DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{
+			LabelSelector: labels.SelectorFromSet(labels.Set{discoveryv1.LabelServiceName: utils.DerivedName(se.Name)}).String(),
+		}); err != nil {
+			// try next time, make sure we clear endpoint slice
+			d := time.Second
+			return &d, err
+		}
+		se.Finalizers = utils.RemoveString(se.Finalizers, known.AppFinalizer)
+		se, err = r.mcsClientset.MulticlusterV1alpha1().ServiceExports(namespace).Update(context.TODO(),
+			se, metav1.UpdateOptions{})
+		if err != nil {
+			d := time.Second
+			return &d, err
+		}
+		klog.Infof("service export %s has been recycled successfully", se.Name)
+		return nil, nil
+	}
+
+	var endpointSliceList *discoveryv1.EndpointSliceList
+	if endpointSliceList, err = r.localk8sClient.DiscoveryV1().EndpointSlices(namespace).List(ctx, metav1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(labels.Set{discoveryv1.LabelServiceName: se.Name}).String(),
+	}); err != nil {
+		if errors.IsNotFound(err) {
+			// service may not exist now, try next time.
+			d := time.Second
+			return &d, err
+		}
+		return nil, err
+	}
+
+	wg := sync.WaitGroup{}
+	var allErrs []error
+	errCh := make(chan error, len(endpointSliceList.Items))
+	for index := range endpointSliceList.Items {
+		wg.Add(1)
+		slice := endpointSliceList.Items[index].DeepCopy()
+		newSlice := r.constructEndpointSlice(slice, se)
+		go func(slice *discoveryv1.EndpointSlice) {
+			defer wg.Done()
+			if err = ApplyEndPointSliceWithRetry(r.parentk8sClient, slice); err != nil {
+				errCh <- err
+				klog.Infof("slice %s sync err: %s", slice.Name, err)
+			}
+		}(newSlice)
+	}
+	wg.Wait()
+	// collect errors
+	close(errCh)
+	for err := range errCh {
+		allErrs = append(allErrs, err)
+	}
+	if len(allErrs) > 0 {
+		reason := utilerrors.NewAggregate(allErrs).Error()
+		msg := fmt.Sprintf("failed to sync endpoint slices of service export %s: %s", klog.KObj(se), reason)
+		klog.ErrorDepth(5, msg)
+		d := time.Second
+		return &d, err
+	}
+	klog.Infof("service export %s has been synced successfully", se.Name)
+	return nil, nil
+}
+
+// constructEndpointSlice construct a new endpoint slice from local slice.
+func (r *reconciler) constructEndpointSlice(slice *discoveryv1.EndpointSlice, se *v1alpha1.ServiceExport) *discoveryv1.EndpointSlice {
+	// mutate slice fields before upload to parent cluster.
+	newSlice := &discoveryv1.EndpointSlice{}
+	newSlice.AddressType = slice.AddressType
+	newSlice.Endpoints = slice.Endpoints
+	newSlice.Ports = slice.Ports
+	newSlice.Labels = make(map[string]string)
+
+	newSlice.GetLabels()[known.LabelServiceName] = se.Name
+	newSlice.GetLabels()[discoveryv1.LabelServiceName] = utils.DerivedName(se.Name)
```

Comment:
We could just use `newSlice.Labels[xxx]` directly.

Reply:
Shall we add a label for service namespace.

Reply:
add subscription namspace: from label: `apps.clusternet.io/subs.namespace`